aten/src/ATen/native/Convolution.cpp

```diff
@@ -851,49 +1005,44 @@ at::Tensor _convolution(
 
   check_shape_forward(input, weight_sizes, bias, params);
 
-  if (input.size(0) == 0 || input.size(1) == 0) {
-    // don't send empty inputs through backends
-    // but need to compute correct output size first and set up history for params
-    std::vector<int64_t> o;
-    if (!params.transposed) {
-      o = conv_output_size(input.sizes(), weight_sizes, params.padding,
-                           params.stride, params.dilation);
-    } else {
-      o = conv_input_size(input.sizes(), weight_sizes, params.padding,
-                          params.output_padding, params.stride, params.dilation,
-                          params.groups);
-    }
-    if (input.size(1) == 0) {
-      o[input_channels_dim] = 0;
-    }
-    if (input_is_mkldnn && weight.is_mkldnn()) {
-      // mkldnn will error on the below 0-dim handling code
-      return empty_mkldnn(
-          o,
-          optTypeMetaToScalarType(input.options().dtype_opt()),
-          input.options().layout_opt(),
-          input.options().device_opt(),
-          input.options().pinned_memory_opt());
-    }
-
-    auto weight_view = at::_unsafe_view(weight, -1);
-    auto out = (input.size(1) == 0) ? (input.view(-1) * weight_view) : (input * weight_view[0]);
-    if (bias.defined()) {
-      out.add_(bias[0]);
-    }
-    return out.view(o);
-  }
-
+  // Expand 1d -> 2d.
   if (k == 3) {
     // avoid accidentally going through NHWC for permuted 3d input.
-    if (!input_is_mkldnn) {
+    if (!input.is_mkldnn()) {
       input = input.contiguous();
     }
     params.view1d_as_2d();
     input = view4d(input);
     weight = view4d(weight);
   }
 
+  // Select appropriate backend to use.
+  ConvBackend backend = select_conv_backend(input, weight, bias, params);
+
+  // Do some additional backend-specific checks.
+  if (backend == ConvBackend::Cudnn || backend == ConvBackend::CudnnTranspose ||
+      backend == ConvBackend::Miopen || backend == ConvBackend::MiopenTranspose) {
+    TORCH_CHECK(input.options().type_equal(weight.options()),
+        "Input type (", input.toString(), ") and weight type (", weight.toString(),
+        ") should be the same");
+    TORCH_CHECK(!bias.defined() || (input.options().type_equal(bias.options())),
+        "Input type (", input.toString(), ") and bias type (", bias.toString(),
+        ") should be the same");
+  }
```

Comment:
nit: it might be nicer if the check were a helper function and then we could put the check in the switch statement below instead of special casing it up here

Reply:
Sounds good, it might also be useful to apply this check more broadly, but I'm punting on that here since it's technically more restrictive.